enf/resource_enf_domain_ratelimit.go

```diff
@@ -0,0 +1,176 @@
+	d.Set("packets_per_second", domainRateLimit.PacketsPerSecond)
+	d.Set("packets_burst_size", domainRateLimit.PacketsBurstSize)
+	d.Set("bytes_per_second", domainRateLimit.BytesPerSecond)
+	d.Set("bytes_burst_size", domainRateLimit.BytesBurstSize)
+
+	return nil
+}
+
+func resourceEnfDomainRateLimitUpdate(d *schema.ResourceData, meta interface{}) error {
+	switch d.Get("type") {
+	case "default":
+		return domainDefaultRateLimitUpdate(d, meta)
+	case "max":
+		return domainMaxRateLimitUpdate(d, meta)
+	default:
+		return fmt.Errorf("Invalid rate limit type %s. The type must be either default or max", d.Get("type"))
+	}
+}
+
+func domainDefaultRateLimitUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*EnfClient).Client
+	domainRateLimitRequest := &enf.DomainRateLimits{
+		PacketsPerSecond: enf.Int(d.Get("packets_per_second").(int)),
+		PacketsBurstSize: enf.Int(d.Get("packets_burst_size").(int)),
+		BytesPerSecond:   enf.Int(d.Get("bytes_per_second").(int)),
+		BytesBurstSize:   enf.Int(d.Get("bytes_burst_size").(int)),
+	}
+	_, _, err := client.Domains.SetDefaultEndpointRateLimits(context.Background(), domainRateLimitRequest, d.Get("domain").(string))
+	if err != nil {
+		return err
+	}
+
+	return resourceEnfDomainRateLimitRead(d, meta)
+}
+
+func domainMaxRateLimitUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*EnfClient).Client
+	domainRateLimitRequest := &enf.DomainRateLimits{
+		PacketsPerSecond: enf.Int(d.Get("packets_per_second").(int)),
+		PacketsBurstSize: enf.Int(d.Get("packets_burst_size").(int)),
+		BytesPerSecond:   enf.Int(d.Get("bytes_per_second").(int)),
+		BytesBurstSize:   enf.Int(d.Get("bytes_burst_size").(int)),
+	}
+	_, _, err := client.Domains.SetMaxDefaultEndpointRateLimits(context.Background(), domainRateLimitRequest, d.Get("domain").(string))
+	if err != nil {
+		return err
+	}
+
+	return resourceEnfDomainRateLimitRead(d, meta)
+}
+
+func resourceEnfDomainRateLimitDelete(d *schema.ResourceData, meta interface{}) error {
+	if d.Get("type") == "max" {
+		return errors.New("The domain max rate limit does not support deletion")
+	}
+
+	// If the type of the rate limit being deleted is default, set the default values to the max rate limit values to simulate deletion.
+	domainMaxRateLimit, _, err := meta.(*EnfClient).Client.Domains.GetMaxDefaultEndpointRateLimits(context.Background(), d.Get("domain").(string))
+	if err != nil {
+		return err
+	}
+
+	domainDefaultRateLimitRequest := &enf.DomainRateLimits{
+		PacketsPerSecond: domainMaxRateLimit.PacketsPerSecond,
+		PacketsBurstSize: domainMaxRateLimit.PacketsBurstSize,
+		BytesPerSecond:   domainMaxRateLimit.BytesPerSecond,
+		BytesBurstSize:   domainMaxRateLimit.BytesBurstSize,
+	}
+
+	_, _, err = meta.(*EnfClient).Client.Domains.SetDefaultEndpointRateLimits(context.Background(), domainDefaultRateLimitRequest, d.Get("domain").(string))
+	if err != nil {
+		return err
+	}
+
+	return resourceEnfDomainRateLimitRead(d, meta)
```

Comment:
Instead of returning the result of `*Read`, I think the delete methods should call `d.SetId("")` and return success?

Reply:
Per the Terraform docs:
`"d.SetId("") is automatically called assuming delete returns no errors"`

I get that this can be confusing though. Do you think I should add a comment describing this behavior?

Reply:
I don't think a comment is necessary.  It's the standard way to do things in Terraform providers, so we don't need to document this case specially.